### PR TITLE
Implement version v1.0.7

### LINKED ISSUE #194: Add playbooks for demand detection and source coverage
## Summary

The current playbook set is strong for:

- dry-run setup / apply decisions
- AI enrichment and AI review
- asset refresh
- breadth-first discovery diagnosis
- recommendation-policy tuning

But there is still a docs gap for the two contributor/operator workflows that sit **under** breadth diagnosis:

1. improving **stack / demand detection** when `discover/output/demand-profile.json` is wrong or weak
2. improving **source coverage / source authoring** when the active discovery universe itself is too narrow

Today `DISCOVERY-BREADTH-PLAYBOOK.md` correctly tells users to distinguish demand detection vs source coverage vs selection vs ranking, but it stops short of a concrete “now fix detection/source coverage” playbook.

## Why this matters

The repo already has substantial implementation and test surface here:

- `src/domains/discovery/demand-profile.ts`
- `src/domains/discovery/demand-signals.ts`
- `src/domains/discovery/detector-signatures.ts`
- `src/domains/discovery/technology-signatures.ts`
- `src/domains/discovery/catalog-selection.ts`
- `src/tests/demand-profile.test.ts`
- `src/tests/detectors.test.ts`
- `src/tests/technology-signatures.test.ts`
- `src/tests/detection-quality.ts`
- `src/tests/detection-fixtures.ts`
- `discover/sources.json`
- `discover/source-packs/*.json`
- `discover/official-skills-indexes.json`
- `discover/official-upstreams.json`

That means contributors can improve these systems today, but the project lacks a first-class playbook that explains the intended workflow.

## Proposed outcome

Add one or both of these playbooks:

### Option A. `DEMAND-DETECTION-PLAYBOOK.md`

Should cover:

- how to diagnose false negatives vs false positives vs weak evidence
- how to inspect `discover/output/demand-profile.json`
- how to reason about detector signatures, technology signatures, and evidence strength
- when to add a data-driven signature vs a focused detector change
- which tests/quality gates to run after changing detection logic
- how to use representative fixtures instead of repo-specific hacks

### Option B. `SOURCE-COVERAGE-PLAYBOOK.md`

Should cover:

- when breadth is a source-registry problem rather than a ranking problem
- how to choose between editing:
  - `discover/sources.json`
  - `discover/source-packs/*.json`
  - `discover/official-skills-indexes.json`
  - `discover/official-upstreams.json`
- how to inspect `source-index.json`, `source-utilization.json`, and selection outputs after source changes
- how to avoid duplicate/noisy/low-value sources
- which validation/tests to run after widening source coverage

## Recommendation

At minimum, add the dedicated **demand detection** playbook. That is the clearest current gap.

If scope allows, add both playbooks so the docs ladder becomes:

- breadth diagnosis ->
- demand/source fix playbook ->
- recommendation policy playbook

## Acceptance criteria

- add a dedicated playbook for demand detection, source coverage, or both
- cross-link it from `DISCOVERY-BREADTH-PLAYBOOK.md` and `README.md`
- include concrete artifact paths, commands, and validation steps
- include an AI-agent prompt section consistent with the existing playbook style
- keep the guidance contributor-oriented and repo-grounded, not generic

### PR DESCRIPTION
**Pull request by ar27111994**
## Summary
- implement the full version v1.0.7 plan on `feat/v1.0.7-version` with separate commits per ticket
- add durable recommendation-policy/runtime improvements, shared CLI/test infrastructure, stronger quality gates, and refreshed operating docs
- prepare the `1.0.7` release metadata and changelog after the full release gate passed

## Included work
- durable base + host recommendation policy overrides and explicit preserve/scale limit override behavior (#204, #192)
- shared recommendation/runtime constants, shared CLI help output helpers, and tighter lint guardrails (#198 and follow-on maintenance)
- reusable built-CLI test harness, dedicated self-hosting coverage, GitHub resilience regressions, and coverage threshold reporting in CI (#199, #194, #195)
- demand/source-coverage/logging playbooks plus roadmap status refreshes for shipped work

## Verification
- `npm run validate:release`
  - `npm run validate:quality`
  - `npm run smoke:cli`
  - `npm run smoke:workspace`
  - `npm run smoke:pack`
- release validation passed after bumping the package to `1.0.7`

Closes #194

## Summary by CodeRabbit

* **New Features**
  * Recommendation limit override modes (`preserve`/`scale`) and durable override layer for base and per-host policies
  * CI coverage reporting and new test commands

* **Documentation**
  * New playbooks: Demand detection and Source coverage; expanded playbooks, logging strategy, and env var override docs
  * Updated changelog and README guidance

* **Bug Fixes**
  * GitHub harvesting resilience with rate-limit fallback and recovery

* **Refactor**
  * Unified CLI help rendering via shared utility

* **Tests**
  * New integration/self-hosting and policy/reporting test coverage

* **Chores**
  * Package version bump and coverage tooling added

[Review Change Stack](https://app.coderabbit.ai/change-stack/ar27111994/agent-harness/pull/206)

<h3>Greptile Summary</h3>

This PR delivers the v1.0.7 release: a durable user-owned recommendation-policy override layer (base + per-host), explicit `preserve`/`scale` limit-override modes with proportional scaling, shared CLI help infrastructure, extracted runtime/selection constants, GitHub resilience regression tests, a self-hosting integration suite, and enforced coverage reporting in CI.

- **Policy override layer** (`src/recommend/policy.ts`, `src/recommend/constants.ts`): `loadRecommendationPolicy` now composes package defaults → user-owned override files → runtime env overrides; `scaleRecommendationHostPolicy` scales related caps proportionally when `mode=scale` is active.
- **Runtime config** (`src/config/runtime.ts`): all inline magic numbers extracted to named constant objects; `buildRecommendationLimitOverrideModes` parses the new `*_LIMIT_MODE` env vars with proper validation.
- **Shared CLI help** (`src/lib/cli-output.ts`, multiple callers): `formatCommandHelp`/`printCommandHelp` replace scattered ad-hoc `console.log` blocks; empty-entries edge case is guarded correctly.

<h3>Confidence Score: 5/5</h3>

The change is safe to merge; the override and scaling logic is well-isolated and covered by targeted tests.

The core policy-merge and scale logic is exercised by new dedicated tests, env isolation is handled correctly in all test helpers, and the runtime override paths have no unguarded edge cases. The one observation about scoring-block replacement is a design trade-off that does not affect existing behavior since no user override files ship with a scoring key.

src/recommend/policy.ts — the mergeRecommendationPolicyBase scoring-replacement semantics are worth a second look before users start writing override files with partial scoring objects.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/recommend/policy.ts | Adds durable user-owned override layer (base + per-host), preserve/scale limit override modes, and the scaleRecommendationHostPolicy helper; mergeRecommendationHostPolicy handles new fields correctly via scalar spread and explicit mergeOptionalUniqueStrings, but mergeRecommendationPolicyBase replaces the entire scoring block rather than merging it. |
| src/config/runtime.ts | Extracts all inline magic numbers into named constant groups and adds buildRecommendationLimitOverrideModes to support the new *_LIMIT_MODE env vars; well-structured, with validation tests confirming invalid modes throw correctly. |
| src/lib/cli-output.ts | New shared formatCommandHelp/printCommandHelp utility; correctly guards against the empty-entries Math.max edge case via an explicit length check. |
| src/recommend/constants.ts | Adds override path constants and extracts selection algorithm weights (preselection multiplier, budget divisor, coverage cap, etc.) as named exports; no logic changes, reduces magic-number drift. |
| src/tests/recommend-policy.test.ts | New test file covering default-policy loading, user-override merging, env-based limit override, scale-mode policy file application, and zero-value skip behavior; all tests use withClearedRecommendationLimitEnv for proper env isolation. |
| src/tests/github-resilience.test.ts | New regression tests for GitHub rate-limit cache fallback and malformed source-health.json recovery; both tests mock globalThis.fetch and clean up correctly via context.after. |
| .github/workflows/quality.yml | Replaces npm test with npm run test:coverage, adds a coverage-summary publishing step (writes markdown to GITHUB_STEP_SUMMARY), and adds npm run test:self-hosting as a dedicated CI step; GITHUB_STEP_SUMMARY is always set in Actions context. |
| src/tests/self-hosting.ts | Integration suite that runs the full discover → recommend pipeline against the repository itself and asserts TypeScript/JavaScript detection, source coverage, and policy output; uses a 300 s timeout to accommodate real I/O. |

</details>

<sub>Reviews (6): Last reviewed commit: ["Clarify coverage exclusion docs"](https://github.com/ar27111994/agent-harness/commit/d11fa507ab63d3e7abbd4a79e88b15e5e7eeb113) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=32184672)</sub>

### CODE REVIEW

**Comment by ar27111994** [comment]
@coderabbitai review

File: .c8rc.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{
+  "reporter": ["text-summary", "lcov", "json-summary"],
+  "check-coverage": true,
+  "statements": 71,
+  "branches": 75,
+  "functions": 74,
+  "lines": 71,
```
Comment (ar27111994) [inline .c8rc.json:7]: I thought that we were doing 100% coverage? Anyway, this is still too low even if 100% isn't feasible yet. Should be at least 90-95% for all of them.

Comment (ar27111994) [reply .c8rc.json:7]: Skipping this one intentionally for this PR. I opened #207 to track the real path to 100% automated coverage, but I did not raise the thresholds here because that would be a larger policy/test-expansion change rather than a minimal review fix.

File: src/lib/state-root.ts [hunk 2]
```diff
@@ -97,6 +103,14 @@ export async function prepareStateRoot(
     }
 
     const destinationPath = resolve(preparedRoot.stateRoot, ...pathSegments);
+    const managedAssetPath = pathSegments.join("/");
+    if (SEEDED_ONCE_STATE_ASSET_PATHS.has(managedAssetPath)) {
+      if (!(await pathExists(destinationPath))) {
+        await copyPath(sourcePath, destinationPath);
+      }
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/lib/state-root.ts:112]: Skipping this one intentionally. Changing state-root seeding here would reverse the preserve-once behavior we explicitly kept in this PR so package recommendation defaults do not silently overwrite user-managed state on later prepares. That is a broader design decision, not a minimal follow-up fix.

File: src/manifest-validation/recommendation.ts [hunk 3]
```diff
@@ -226,6 +227,31 @@ export function assertRecommendationReport(
...
+    assertLiteral(
+      summaryRecord.recommendationLimitOverrideMode,
+      ["preserve", "scale"],
+      `${context}.hostSummaries.${host}.recommendationLimitOverrideMode`,
+    );
+    assertLiteral(
+      summaryRecord.recommendationLimitOverrideModeSource,
+      ["policy", "env"],
+      `${context}.hostSummaries.${host}.recommendationLimitOverrideModeSource`,
+    );
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/manifest-validation/recommendation.ts]: Addressed in `e22f131`. Recommendation report validation now backfills missing legacy host-summary override metadata so older persisted reports do not fail validation on upgrade.

File: src/tests/recommend-policy.test.ts [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+void test("recommendation policy loads package defaults by default", async () => {
+  await withPolicyWorkspace(async (projectRoot) => {
+    const policy = await loadRecommendationPolicy(projectRoot);
+
+    assert.equal(policy.hosts["copilot-vscode"].recommendationLimit, 240);
+    assert.equal(
+      policy.hosts["copilot-vscode"].recommendationLimitOverrideMode,
+      "preserve",
+    );
+    assert.equal(policy.hosts["copilot-vscode"].maxPerAssetKind.skill, 72);
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-policy.test.ts]: Addressed in `e22f131`. The recommendation-policy tests now clear and restore the relevant recommendation-limit env vars so they do not depend on the parent process environment.

File: src/tests/recommend-report.test.ts [hunk 5]
```diff
@@ -256,6 +265,73 @@ void test("recommendation reports expose env-sourced recommendation limits", asy
...
+  const previousLimitEnvValue =
+    process.env.AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT;
+  const previousModeEnvValue =
+    process.env.AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT_MODE;
+  process.env.AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT = "120";
+  process.env.AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT_MODE = "scale";
+
+  try {
+    loadRuntimeConfig(process.env);
+    const policy = await loadRecommendationPolicy(process.cwd());
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-report.test.ts]: Addressed in `e22f131`. I removed the misleading no-op `loadRuntimeConfig(process.env)` call from the recommend-report tests.

File: src/types/recommendation.ts [hunk 6]
```diff
@@ -91,13 +91,21 @@ export interface RecommendationPolicyPresetRefs {
   targetConcerns?: string[];
 }
 
+/**
+ * Describes recommendation host policy data exchanged by the lifecycle pipeline.
+ */
+export type RecommendationLimitOverrideMode = "preserve" | "scale";
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/types/recommendation.ts:97]: Addressed in `e22f131`. The exported doc comment now correctly describes `RecommendationLimitOverrideMode` rather than host policy data.

File: src/tests/recommend-report.test.ts [hunk 7]
```diff
@@ -245,6 +245,15 @@ void test("recommendation reports expose env-sourced recommendation limits", asy
...
       "AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT",
     );
+    assert.equal(
+      report.hostSummaries["copilot-vscode"].recommendationLimitOverrideMode,
+      "preserve",
+    );
+    assert.equal(
+      report.hostSummaries["copilot-vscode"]
+        .recommendationLimitOverrideModeSource,
+      "policy",
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-report.test.ts:284]: Addressed in `e22f131`. The recommend-report tests now save and restore the override-mode env var as well, so the assertions are hermetic.

File: package.json [hunk 8]
```diff
@@ -82,6 +82,8 @@
     "check": "npm run typecheck",
     "check:version-sync": "node ./scripts/check-version-sync.mjs",
     "test": "node --test dist/tests/*.test.js scripts/tests/*.test.mjs",
+    "test:coverage": "c8 node --test dist/tests/*.test.js scripts/tests/*.test.mjs",
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply package.json:85]: Addressed in `e22f131`. `coverage/` is now ignored so local `test:coverage` runs do not leave untracked coverage artifacts.

File: src/tests/recommend-policy.test.ts [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+    assert.equal(
+      printedPolicy.runtimeOverrides.recommendationLimitOverrideMode,
+      "scale",
+    );
+    assert.equal(
+      printedPolicy.runtimeOverrides.recommendationLimitOverrideModeSource,
+      "policy",
+    );
+    assert.equal(printedPolicy.runtimeOverrides.scalingApplied, false);
+  });
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-policy.test.ts:118]: Good catch — this one was still valid. Fixed in `b349f50` by wrapping the test in `withClearedRecommendationLimitEnv(...)`, and I revalidated with `npm run build && node --test dist/tests/recommend-policy.test.js`.

File: src/tests/recommend-policy.test.ts [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+      {
+        schemaVersion: 1,
+        host: "copilot-vscode",
+        policy: {
+          maxPerAssetKind: {
+            skill: 0,
+          },
+        },
+      },
+    );
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-policy.test.ts]: Good catch — this one was still valid. Fixed in `99f721f` by moving the `writePolicyFile(...)` call inside the `try` so the env restoration in `finally` still runs if the file write throws. Revalidated with `npm run build && node --test dist/tests/recommend-policy.test.js`.

File: FUTURE-IMPROVEMENTS.md [hunk 11]
```diff
@@ -265,27 +297,30 @@ The suggestions below are refinements, enhancements, and operational improvement
 
 ## 12. Test suite and validation harness
 
+**Status:** Implemented / Partial
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply FUTURE-IMPROVEMENTS.md]: Fixed in `99f721f`. I changed this status label to `Partial` so it matches the three-label taxonomy introduced at the top of the document.

File: .c8rc.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (ar27111994) [reply .c8rc.json:7]: Skipping this one intentionally for this PR. I agree the current gate is not true all-files coverage, but enabling `c8` all-files instrumentation here would materially change the semantics of the coverage gate and likely require a broader coverage expansion pass. I captured that larger direction in #207 instead of silently changing the gate mid-PR.

File: src/tests/self-hosting.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+      timeout: SELF_HOSTING_TIMEOUT_MS,
+      args: ["discover", "sources"],
+    });
+    await runBuiltCli({
+      cwd: repositoryRoot,
+      env,
+      stateRoot,
+      timeout: SELF_HOSTING_TIMEOUT_MS,
+      args: ["discover", "catalog"],
+    });
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/self-hosting.ts:58]: Skipping this one intentionally for this PR. This self-hosting suite is an explicit slow live integration test, and keeping it real rather than mocked is part of what it is meant to verify. Making it fully deterministic would require a broader redesign of the fixture/source strategy rather than a minimal review fix.

File: src/tests/recommend-policy.test.ts [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+        delete process.env
+          .AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT_MODE;
+      } else {
+        process.env.AGENT_HARNESS_COPILOT_VSCODE_RECOMMENDATION_LIMIT_MODE =
+          previousModeEnvValue;
+      }
+      clearRuntimeConfigForTests();
+    }
+  });
+});
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-policy.test.ts:178]: Good catch — this was still valid. Fixed in `48cd5a9` by switching this test to `withClearedRecommendationLimitEnv(...)` instead of manual env save/restore, and revalidated with `npm run build && node --test dist/tests/recommend-policy.test.js`.

File: src/tests/recommend-policy.test.ts [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/recommend-policy.test.ts:233]: Fixed in `48cd5a9` as well. This test now also uses `withClearedRecommendationLimitEnv(...)` instead of manual env management, so both policy-print env-sensitive tests are consistent and hermetic.

File: README.md [hunk 15]
```diff
@@ -1370,9 +1389,12 @@ Before pushing changes, run at least:
 ```bash
 npm run validate
 npm run build
-npm test
+npm run test:coverage
+npm run test:self-hosting
 ```
 
+Coverage is enforced through `npm run test:coverage` using the checked-in `.c8rc.json` policy. The current gate targets shipped runtime modules rather than generated types, test harnesses, or thin top-level command shims, and it fails CI if statements/lines drop below 71%, branches below 75%, or functions below 74%.
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply README.md]: Good catch — this was a docs mismatch, not a config change. Fixed in `d11fa50` by tightening the README wording so it refers to the exclusions as selected built command-entry outputs listed in `.c8rc.json`, rather than calling them thin shims.
